In [1]:
import numpy as np
import pandas as pd
import pysubgroup as ps
from mlxtend.frequent_patterns import apriori, association_rules

In [24]:
df_original = pd.read_csv('./Data/tournaments.csv')
df = df_original.copy()

#### Tratando dataframe

In [4]:
df.replace(np.nan, 0, inplace=True)
df.tail()

id_card         name_card  amount_card  price_card energy_type_card  \
114286  GRI130a  Rescue Stretcher            2        0.72                0   
114287   GRI119        Aqua Patch            2        0.27                0   
114288   FLF092           Pal Pad            1        0.68                0   
114289   GRI121       Choice Band            2        0.20                0   
114290   GRI120     Brooklet Hill            2        0.32                0   

       type_card  combo_type_id   combo_type_name  id_player  name_player  \
114286   Trainer           47.0  Zoroark Greninja       1900  Jin Woo Lee   
114287   Trainer           47.0  Zoroark Greninja       1900  Jin Woo Lee   
114288   Trainer           47.0  Zoroark Greninja       1900  Jin Woo Lee   
114289   Trainer           47.0  Zoroark Greninja       1900  Jin Woo Lee   
114290   Trainer           47.0  Zoroark Greninja       1900  Jin Woo Lee   

        ... region_tournament  country_tournament  year_tournament  \
114286  ...             AS-OC               Korea             2018   
114287  ...             AS-OC               Korea             2018   
114288  ...             AS-OC               Korea             2018   
114289  ...             AS-OC               Korea             2018   
114290  ...             AS-OC               Korea             2018   

        month_tournament day_tournament valid_rotation_at_tournament  \
114286                12             22                            0   
114287                12             22                            0   
114288                12             22                            0   
114289                12             22                            0   
114290                12             22                            0   

        rotation_name year_begin  month_begin  day_begin  
114286  standard_2019       2018            8         28  
114287  standard_2019       2018            8         28  
114288  standard_2019       2018            8         28  
114289  standard_2019       2018            8         28  
114290  standard_2019       2018            8         28  

[5 rows x 26 columns]

#### Tentaremos encontrar subgrupos frequentes de cartas que aparecem em torneios diferentes de batalha de cartas pokemon no ano de 2019

- Definindo todas as regioes do dataset

In [6]:
df = df[df['region_tournament'] != 0]

# eliminando cartas do tipo treinador
df = df[df['type_card'] != 'Trainer']
regions = df['region_tournament'].unique()

regions

array(['SA', 'EU', 'AS-OC', 'JP'], dtype=object)

- Definindo treinadores com suas cartas para cada regiao e realizando apriori para descobrir padroes de cartas utilizadas por jogadores diferentes

In [21]:
# Agrupando cartas de cada jogador no dataset
# Dessa forma, separamos os dados no formato das transações utilizadas no Apriori
def agrupa_jogador_cartas(df):
    dict_jogador_cartas = {}
    for _, row in df.iterrows():
        if row['name_player'] not in dict_jogador_cartas:
            dict_jogador_cartas[row['name_player']] = []
        dict_jogador_cartas[row['name_player']].append(row['name_card'])
    return len(dict_jogador_cartas), dict_jogador_cartas

# Fazendo filtro por região somente no ano de 2019
info_por_regiao = {}
for region in regions:
    print(f"region: {region}")
    df_region = df[(df['region_tournament'] == region) & (df['year_tournament'] == 2019)]
    jogadores = df_region['name_player'].unique()
    cartas = df_region['name_card'].unique()

    quantidade_jogadores, dict_jogador_cartas = agrupa_jogador_cartas(df_region)

    # Cria um dataframe que as colunas sao todas as cartas existentes no dataset
    df_jogadores_cartas = pd.DataFrame(columns=cartas)

    for jogador in jogadores:
        cartas_jogador = dict_jogador_cartas[jogador]
        linha = []
        # Construo matriz de incidencia -> True se o jogador possui a carta, False caso contrario
        for carta in cartas:
            if carta in cartas_jogador:
                linha.append(True)
            else:
                linha.append(False)
        df_jogadores_cartas.loc[len(df_jogadores_cartas)] = linha

    # Apriori

    # Suporte - O quanto a carta aparece no total de cartas usadas
    frequent_itemsets = apriori(df_jogadores_cartas, min_support=0.22, use_colnames=True) # use_colnames -> nome das cartas ao inves de indices
    frequent_itemsets.sort_values(by='support', ascending=False, inplace=True)

    # Regras: Se compro A (antecedente) -> compro B (consequente) com confianca de X%
    rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.4) # confiança das regras geradas -> 40%
    # Lift -> Se a regra é relevante ou não (se é maior que 1, é relevante)
    # Ignoro algumas colunas para facilitar a visualização
    rules = rules.sort_values(by='lift', ascending=False).drop(['antecedent support', 'consequent support', 'leverage', 'conviction', 'zhangs_metric'], axis=1)
    
    info_por_regiao[region] = {'itemsets': frequent_itemsets[:3], 'rules': rules[:3]}

    print(f"Total de jogadores: {quantidade_jogadores}")
    print("Itemsets frequentes")
    print(frequent_itemsets)
    print("Regras")
    print(rules)
    print("\n\n")

region: SA
Total de jogadores: 73
Itemsets frequentes
    support                                    itemsets
0  0.589041                                   (Jirachi)
1  0.493151                                (Dedenne-GX)
2  0.410959              (Mega Lopunny & Jigglypuff-GX)
6  0.342466                       (Dedenne-GX, Jirachi)
3  0.301370               (Arceus & Dialga & Palkia-GX)
7  0.287671  (Mega Lopunny & Jigglypuff-GX, Dedenne-GX)
4  0.246575                                 (Keldeo-GX)
5  0.246575                                 (Cryogonal)
8  0.246575    (Arceus & Dialga & Palkia-GX, Cryogonal)
Regras
                      antecedents                     consequents   support  \
4   (Arceus & Dialga & Palkia-GX)                     (Cryogonal)  0.246575   
5                     (Cryogonal)   (Arceus & Dialga & Palkia-GX)  0.246575   
2  (Mega Lopunny & Jigglypuff-GX)                    (Dedenne-GX)  0.287671   
3                    (Dedenne-GX)  (Mega Lopunny & Jigglypuff-G

### ANALISAR --> PQ AS REGRAS SÃO REPETIDAS?
EX: ANTECEDENTE A - CONSEQUENTE B
ANTECEDENTE B - CONSEQUENTE A

TEM COMO RETIRAR ISSO?

- Analisando os resultados mais relevantes de cada região:

In [23]:
for region in regions:
    print(f"Região: {region}")
    print("Itemsets frequentes")
    print(info_por_regiao[region]['itemsets'])
    print("Regras")
    print(info_por_regiao[region]['rules'])
    print("\n\n")

Região: SA
Itemsets frequentes
    support                        itemsets
0  0.589041                       (Jirachi)
1  0.493151                    (Dedenne-GX)
2  0.410959  (Mega Lopunny & Jigglypuff-GX)
Regras
                      antecedents                    consequents   support  \
4   (Arceus & Dialga & Palkia-GX)                    (Cryogonal)  0.246575   
5                     (Cryogonal)  (Arceus & Dialga & Palkia-GX)  0.246575   
2  (Mega Lopunny & Jigglypuff-GX)                   (Dedenne-GX)  0.287671   

   confidence      lift  
4    0.818182  3.318182  
5    1.000000  3.318182  
2    0.700000  1.419444  



Região: EU
Itemsets frequentes
    support       itemsets
5  0.650602  (Tapu Koko ♢)
0  0.590361      (Jirachi)
7  0.590361    (Marshadow)
Regras
                     antecedents                  consequents   support  \
323  (Tapu Lele-GX, Tapu Koko ♢)      (Marshadow, Zeraora-GX)  0.240964   
324      (Marshadow, Zeraora-GX)  (Tapu Lele-GX, Tapu Koko ♢)  0.24096

In [33]:
print("Media de preço das cartas em todos os torneios em todos os anos" , df['price_card'].mean())
print("Media de preço das cartas em todos os torneios no ano de 2019" , df[(df['year_tournament'] == 2019)]['price_card'].mean())
print("Preço da carta Jirachi em 2019" , df[(df['year_tournament'] == 2019) & (df['name_card'] == 'Jirachi')]['price_card'].mean())

Media de preço das cartas em todos os torneios em todos os anos 1.9965210079474869
Media de preço das cartas em todos os torneios no ano de 2019 2.9583872688477952
Preço da carta Jirachi em 2019 6.5


In [34]:
print("Media de quantidade máxima de cartas iguais em todos os torneios em todos os anos" , df['amount_card'].mean())
print("Media de preço das cartas em todos os torneios no ano de 2019" , df[(df['year_tournament'] == 2019)]['amount_card'].mean())
print("Preço da carta Jirachi em 2019" , df[(df['year_tournament'] == 2019) & (df['name_card'] == 'Jirachi')]['amount_card'].mean())

Media de quantidade máxima de cartas iguais em todos os torneios em todos os anos 2.1508692722961564
Media de preço das cartas em todos os torneios no ano de 2019 2.173510396645116
Preço da carta Jirachi em 2019 3.189922480620155


### SA
Jirachi, Dedenne-GX, Mega Lopunny & Jigglypuff-GX
Regras Arceus & Dialga & Palkia-GX e Cryogonal

### EU
Tapu Koko ♢, Jirachi, Marshdown
Regras envolvendo Tapu Lele-GX, Tapu Koko ♢, Marshdown e Zeraora-GX

### AS-OC
Jirachi, Tapu Lele-GX, Marshdown
Destaque para regras enolvendo Tapu Koko ♢, Jirachi, Zapdos e Absol

### JP
Jirachi, Dedenne-GX, Marshdow (destaque para regras envolvendo essa carta)

Percebe-se que a carta Jirachi foi almplamente utilizadas em torneios de 2019 em diferentes regiões ao redor do mundo. Jirachi é uma carta com um alto custo em comparação a média dos valores nos torneios (média de 2.95 contra custo de 6.5 da carta Jirachi em 2019), porém ao mesmo tempo, pode ser mais utilizada em um mesmo deck (3 vezes em comparação a média de 2.17 naquele ano)


- Analise para combos de cartas

In [90]:
df = df_original.copy()

df = df[(df['year_tournament'] == 2019)]


df.replace(np.nan, 0, inplace=True)
df.tail()

id_card             name_card  amount_card  price_card  \
113994   UNB170        Energy Spinner            1        0.13   
113995  UPR122a          Escape Board            2        0.00   
113996  GRI121a           Choice Band            2       19.74   
113997   CES143  Shrine of Punishment            2        0.33   
113998   LOT191    Thunder Mountain ♢            1        1.07   

       energy_type_card type_card  combo_type_id combo_type_name  id_player  \
113994                0   Trainer           84.0          Zapdos       1889   
113995                0   Trainer           84.0          Zapdos       1889   
113996                0   Trainer           84.0          Zapdos       1889   
113997                0   Trainer           84.0          Zapdos       1889   
113998                0   Trainer           84.0          Zapdos       1889   

       name_player  ... region_tournament  country_tournament  \
113994   Sumin Lim  ...             AS-OC               Korea   
113995   Sumin Lim  ...             AS-OC               Korea   
113996   Sumin Lim  ...             AS-OC               Korea   
113997   Sumin Lim  ...             AS-OC               Korea   
113998   Sumin Lim  ...             AS-OC               Korea   

        year_tournament  month_tournament day_tournament  \
113994             2019                 2             16   
113995             2019                 2             16   
113996             2019                 2             16   
113997             2019                 2             16   
113998             2019                 2             16   

       valid_rotation_at_tournament  rotation_name year_begin  month_begin  \
113994                      SUM-TEU  standard_2019       2018            8   
113995                      SUM-TEU  standard_2019       2018            8   
113996                      SUM-TEU  standard_2019       2018            8   
113997                      SUM-TEU  standard_2019       2018            8   
113998                      SUM-TEU  standard_2019       2018            8   

        day_begin  
113994         28  
113995         28  
113996         28  
113997         28  
113998         28  

[5 rows x 26 columns]

In [95]:
# If 'presence' column doesn't exist, create it for the BinaryTarget
df['presence'] = (df['name_card'] == 'Jirachi').astype(int)

In [91]:
len(df)

11446

In [99]:
# Define the target for subgroup discovery
target = ps.BinaryTarget('presence', 1)

# Define the search space
combo_selectors = [ps.EqualitySelector('combo_type_name', value) for value in df['combo_type_name'].unique()]
ranking_selector = [ps.EqualitySelector('ranking_player_tournament', value) for value in df['ranking_player_tournament'].unique()]
category_selectors = [ps.EqualitySelector('category_tournament', value) for value in df['category_tournament'].unique()]
region_selectors = [ps.EqualitySelector('region_tournament', value) for value in df['region_tournament'].unique()]
search_space = combo_selectors +  ranking_selector + category_selectors + region_selectors

In [83]:
search_space

[combo_type_name=='Reshiram & Charizard Fire Box',
 combo_type_name=='Mewtwo & Mew',
 combo_type_name=='ADP Keldeo',
 combo_type_name=='Blacephalon Naganadel',
 combo_type_name=='Florges Dolls',
 combo_type_name=='Garchomp & Giratina',
 combo_type_name=='Green's ADP',
 combo_type_name=='Blacephalon Pidgeotto',
 combo_type_name=='Pidgeotto Control',
 combo_type_name=='Gardevoir & Sylveon',
 combo_type_name=='Naganadel & Guzzlord',
 combo_type_name=='Charizard & Braixen',
 combo_type_name=='Malamar Giratina',
 combo_type_name=='Quagsire Silvally',
 combo_type_name=='Pikachu & Zekrom',
 combo_type_name=='Mewtwo & Mew Fire Box',
 combo_type_name=='Quagsire Naganadel',
 combo_type_name=='Charizard',
 combo_type_name=='Blacephalon Silvally',
 combo_type_name=='Zoroark Dewgong',
 combo_type_name=='Blacephalon',
 combo_type_name=='Pikachu & Zekrom Zapdos',
 combo_type_name=='Ultra Malamar ',
 combo_type_name=='Zapdos',
 combo_type_name=='Lucario & Melmetal Vileplume',
 combo_type_name=='Spirit

In [102]:

task = ps.SubgroupDiscoveryTask(df, target, search_space, result_set_size=15, depth=5, qf=ps.WRAccQF())

# Run the SD-Map algorithm
result = ps.SimpleDFS().execute(task)


In [103]:
# Display the results
for row in result.to_dataframe().itertuples():
    print(row)

Pandas(Index=0, quality=0.001806979198588307, subgroup=(combo_type_name=='Ultra Malamar '), size_sg=857, size_dataset=11446, positives_sg=40, positives_dataset=258, size_complement=10589, relative_size_sg=0.07487331818976062, relative_size_complement=0.9251266818102394, coverage_sg=0.15503875968992248, coverage_complement=0.8449612403100775, target_share_sg=0.046674445740956826, target_share_complement=0.02058740202096515, target_share_dataset=0.022540625546042286, lift=2.0706810308177976)
Pandas(Index=1, quality=0.0017953465801277201, subgroup=(combo_type_name=='Zapdos'), size_sg=996, size_dataset=11446, positives_sg=43, positives_dataset=258, size_complement=10450, relative_size_sg=0.0870172986196051, relative_size_complement=0.912982701380395, coverage_sg=0.16666666666666666, coverage_complement=0.8333333333333334, target_share_sg=0.04317269076305221, target_share_complement=0.02057416267942584, target_share_dataset=0.022540625546042286, lift=1.9153279785809907)
Pandas(Index=2, qual

# A ideia é achar subgrupos ligados a carta Jirachi - Mas pela qualidade não deu muito certo